### DISCUSSION TITLE
Necessary kernel config options

### DISCUSSION THREAD

**Discussion by kcning** [evilsocket/opensnitch#978] [Q&A] [answered]
Hello, thank you so much for the various improvements in the latest release.

Opensnitch requires various kernel config options to function. For packaging purposes, it would be nice to isolate
them into groups. I checked some [doc](https://github.com/evilsocket/opensnitch/tree/master/ebpf_prog) and issues (https://github.com/evilsocket/opensnitch/issues/774, https://github.com/evilsocket/opensnitch/issues/976). It seems that

1. needed for intercepting connections, regardless of which firewall is used: 
        NETFILTER_NETLINK_ACCT
        NETFILTER_NETLINK_QUEUE
        NF_CONNTRACK
        NF_CT_NETLINK
        
2. needed to use iptables as firewall:
        NETFILTER_XT_MATCH_CONNTRACK
        NETFILTER_XT_TARGET_NFQUEUE
        
3. needed to use nftables as firewall:
        NFT_CT
        NFT_QUEUE

4. needed for ebpf monitor:
    CGROUP_BPF
    BPF_EVENTS
    FTRACE_SYSCALLS
    KPROBES_ON_FTRACE
    KPROBE_EVENTS
    UPROBE_EVENTS

Do I miss and misunderstand something? And what would be the options needed for proc and audit monitoring methods?

**Comment by gustavo-iniguez-goya** [comment]
hi @kcning !

Yes, those are the needed configs. Depending on some configurations there coud be others, but I think that for most kernels those are pretty much it (based on what I've tested, from 4.4 to 6.3).

I added a `-check-requirements` option to `opensnitchd` that dumps the requisites by module/category:
https://github.com/evilsocket/opensnitch/blob/8240b7bfdac654ccc36816f205c3e1e5fa84de82/daemon/core/system.go#L64

Not all are mandatory though. For example UPROBE_EVENTS and *SYSCALLS are not strictly needed, but we'd lose some functionality otherwise.

>  And what would be the options needed for proc and audit monitoring methods?

`CONFIG_PROC_FS=y` for proc,  and for audit (I haven't tested kernels with audit disabled, so ...):
```
CONFIG_AUDIT=y
CONFIG_HAVE_ARCH_AUDITSYSCALL=y
CONFIG_AUDITSYSCALL=y
CONFIG_AUDIT_ARCH=y
```

**Reply by gustavo-iniguez-goya** [reply]
oh, by the way, for all the monitor methods these config options are also needed:
```
CONFIG_INET_DIAG=m
CONFIG_INET_TCP_DIAG=m
CONFIG_INET_UDP_DIAG=m
CONFIG_INET_RAW_DIAG=m
```

not strictly, but they're very useful.

`CONFIG_INET_DIAG_DESTROY=y` -> this is mandatory to use the `reject` feature.

**Reply by kcning** [reply]
Thank you for the quick reply! Nice that proc and audit methods require only the mininum to work.
And although it's a very unlikely scenario, do ebpf and audit methods work with kernels with CONFIG_PROC_FS=n?

**Reply by gustavo-iniguez-goya** [reply, marked as answer]
No, right now procfs is used from both ebpf and audit.

**Reply by kcning** [reply]
I have a rough idea now. Thank you.
